addons/iap/models/iap_account.py

```diff
@@ -74,32 +161,49 @@ def get(self, service_name, force_create=True):
         return accounts[0]
 
     @api.model
-    def get_credits_url(self, service_name, base_url='', credit=0, trial=False):
+    def get_credits_url(self, service_name, base_url='', credit=0, trial=False, account_token=False, get_service_from_token=False):
         """ Called notably by ajax crash manager, buy more widget, partner_autocomplete, sanilmail. """
         dbuuid = self.env['ir.config_parameter'].sudo().get_param('database.uuid')
         if not base_url:
             endpoint = iap_tools.iap_get_endpoint(self.env)
             route = '/iap/1/credit'
             base_url = endpoint + route
-        account_token = self.get(service_name).account_token
+        if not account_token:
+            account_token = self.get(service_name).account_token
         d = {
             'dbuuid': dbuuid,
             'service_name': service_name,
             'account_token': account_token,
             'credit': credit,
+            'get_service_from_token': get_service_from_token,
         }
         if trial:
             d.update({'trial': trial})
         return '%s?%s' % (base_url, werkzeug.urls.url_encode(d))
 
+    def action_buy_credits(self):
+        for account in self:
+            return {
+                'type': 'ir.actions.act_url',
+                'url': self.env['iap.account'].get_credits_url(
+                    account_token=account.account_token,
+                    service_name=account.service_name,
```

Comment:
If the `service_name` is currently the account name, it won't work, right ?
It should be `iap_service_name`.